From dc3e08cb3f0df9d7f6b2292f76ef061a73e079c1 Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <thomas.langer@intel.com>
Date: Fri, 10 Jan 2020 19:45:28 +0100
Subject: [PATCH] PONRTSYS-6161: Support a minimal address for kASLR

With this it is possible the specify an address which is below the
default load address. If kASLR is disabled (on cmdline), it will still
relocate the kernel to this location.
---
 arch/mips/Kconfig           | 10 +++++++
 arch/mips/kernel/relocate.c | 71 +++++++++++++++++++++++----------------------
 2 files changed, 46 insertions(+), 35 deletions(-)

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index c287b55bf1ca..5d7abcdc4027 100755
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -2697,6 +2697,16 @@ config RANDOMIZE_BASE_MAX_OFFSET
 	  This is limited by the size of KSEG0, 256Mb on 32-bit or 1Gb with
 	  EVA or 64-bit. The default is 16Mb.
 
+config RANDOMIZE_BASE_MIN_ADDR
+	hex "Minimal kASLR address" if EXPERT
+	depends on RANDOMIZE_BASE
+	default "0x00000000"
+	---help---
+	  When kASLR is active, this can be used to configure the minimal
+	  address for the relocation.
+	  If this is not 0, it will be used as base address instead of the
+	  default load address of the kernel.
+
 config NODES_SHIFT
 	int
 	default "6"
diff --git a/arch/mips/kernel/relocate.c b/arch/mips/kernel/relocate.c
index 99dd3b08f331..f2cf9ab1cf51 100644
--- a/arch/mips/kernel/relocate.c
+++ b/arch/mips/kernel/relocate.c
@@ -242,13 +242,39 @@ static inline __init bool kaslr_disabled(void)
 	return false;
 }
 
+static inline int __init relocation_addr_valid(void *loc_new)
+{
+	unsigned long kernel_length;
+
+	if ((unsigned long)loc_new & 0x0000ffff) {
+		/* Inappropriately aligned new location */
+		return 0;
+	}
+	kernel_length = (long)_end - (long)(&_text);
+
+	/* enough space before kernel */
+	if ((unsigned long)loc_new + kernel_length < (unsigned long)&_text)
+		return 1;
+
+	/* behind the original kernel */
+	if ((unsigned long)loc_new > (unsigned long)&_end)
+		return 1;
+
+	/* New location overlaps original kernel */
+	pr_err("locations overlap for 0x%p\n", loc_new);
+	return 0;
+}
+
 static inline void __init *determine_relocation_address(void)
 {
 	/* Choose a new address for the kernel */
 	unsigned long kernel_length;
 	void *dest = &_text;
-	unsigned long offset;
-	int count = 100;
+	unsigned long offset = 0;
+	int count = 10;
+
+	if (CONFIG_RANDOMIZE_BASE_MIN_ADDR != 0)
+		dest = (void *)CONFIG_RANDOMIZE_BASE_MIN_ADDR;
 
 	if (kaslr_disabled())
 		return dest;
@@ -258,20 +284,19 @@ static inline void __init *determine_relocation_address(void)
 	do {
 		offset = get_random_boot() << 16;
 		offset &= (CONFIG_RANDOMIZE_BASE_MAX_OFFSET - 1);
-		if (offset < kernel_length)
-			offset += ALIGN(kernel_length, 0xffff);
 
-		if (offset + kernel_length <= CONFIG_RANDOMIZE_BASE_MAX_OFFSET)
-			break;
+		if (relocation_addr_valid(RELOCATED(dest)))
+			return RELOCATED(dest);
+
 		pr_debug("Relocation by 0x%lX not possible\n", offset);
 	} while (--count);
 
-	if (!count) {
+	if (dest == &_text)
 		pr_err("Relocation not possible, did not find a working offset\n");
-		offset = 0;
-	}
+	else
+		pr_err("Relocation to fixed offset\n");
 
-	return RELOCATED(dest);
+	return dest;
 }
 
 #else
@@ -287,28 +312,6 @@ static inline void __init *determine_relocation_address(void)
 
 #endif
 
-static inline int __init relocation_addr_valid(void *loc_new)
-{
-	unsigned long kernel_length;
-
-	if ((unsigned long)loc_new & 0x0000ffff) {
-		/* Inappropriately aligned new location */
-		return 0;
-	}
-	kernel_length = (long)_end - (long)(&_text);
-
-	/* enough space before kernel */
-	if ((unsigned long)loc_new + kernel_length < (unsigned long)&_text)
-		return 1;
-
-	/* behind the original kernel */
-	if ((unsigned long)loc_new > (unsigned long)&_end)
-		return 1;
-
-	/* New location overlaps original kernel */
-	return 0;
-}
-
 void *__init relocate_kernel(void)
 {
 	void *loc_new;
@@ -335,9 +338,7 @@ void *__init relocate_kernel(void)
 
 	loc_new = determine_relocation_address();
 
-	/* Sanity check relocation address */
-	if (relocation_addr_valid(loc_new))
-		offset = (unsigned long)loc_new - (unsigned long)(&_text);
+	offset = (unsigned long)loc_new - (unsigned long)(&_text);
 
 	/* Reset the command line now so we don't end up with a duplicate */
 	arcs_cmdline[0] = '\0';
